src/protocol/sort/compose.rs

```diff
@@ -0,0 +1,127 @@
+    /// 4. Revealed permutation is applied locally on another permutation shares (rho)
+    /// 5. Unshuffle the permutation with random permutations to negate the effect of shuffling
+    #[allow(dead_code)]
+    pub async fn execute(&mut self, ctx: ProtocolContext<'_, F>) -> Result<(), BoxError> {
+        let mut random_permutations =
+            get_two_of_three_random_permutations(self.rho.len(), &ctx.prss());
+        let mut random_permutations_copy = random_permutations.clone();
+
+        let mut shuffle_sigma = Shuffle::new(self.sigma);
+        shuffle_sigma
+            .execute(ctx.narrow(&ShuffleSigma), &mut random_permutations)
+            .await?;
+
+        let mut perms = reveal_a_permutation(ctx.narrow(&RevealPermutation), self.sigma).await?;
+
+        apply_inv(&mut perms, &mut self.rho);
+
+        let mut unshuffle_rho = Shuffle::new(self.rho);
+        unshuffle_rho
+            .execute_unshuffle(ctx.narrow(&UnshuffleRho), &mut random_permutations_copy)
+            .await?;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use permutation::Permutation;
+    use rand::seq::SliceRandom;
+    use tokio::try_join;
+
+    use crate::{
+        protocol::{
+            sort::{apply::apply_inv, compose::Compose},
+            QueryId,
+        },
+        test_fixture::{
+            generate_shares, make_contexts, make_world, validate_result_from_shares, TestWorld,
+        },
+    };
+
+    #[tokio::test]
+    pub async fn compose() {
+        const BATCHSIZE: usize = 25;
+        for _ in 0..10 {
+            let mut rng_sigma = rand::thread_rng();
+            let mut rng_rho = rand::thread_rng();
+
+            let mut sigma: Vec<usize> = (0..BATCHSIZE).collect();
+            sigma.shuffle(&mut rng_sigma);
+
+            let sigma_u128: Vec<u128> = sigma.iter().map(|x| *x as u128).collect();
+
+            let mut rho: Vec<usize> = (0..BATCHSIZE).collect();
+            rho.shuffle(&mut rng_rho);
+            let rho_u128: Vec<u128> = rho.iter().map(|x| *x as u128).collect();
+
+            let mut rho_composed = rho_u128.clone();
+
+            let mut sigma_copy = Permutation::oneline(sigma.clone());
+            apply_inv(&mut sigma_copy, &mut rho_composed);
+
+            let mut sigma_shares = generate_shares(sigma_u128);
+            let mut rho_shares = generate_shares(rho_u128);
+            let world: TestWorld = make_world(QueryId);
+            let [ctx0, ctx1, ctx2] = make_contexts(&world);
+
+            let mut compose0 = Compose::new(&mut sigma_shares.0, &mut rho_shares.0);
+            let mut compose1 = Compose::new(&mut sigma_shares.1, &mut rho_shares.1);
+            let mut compose2 = Compose::new(&mut sigma_shares.2, &mut rho_shares.2);
+
+            let h0_future = compose0.execute(ctx0);
+            let h1_future = compose1.execute(ctx1);
+            let h2_future = compose2.execute(ctx2);
+
+            try_join!(h0_future, h1_future, h2_future).unwrap();
```

Comment:
If you make the test return a `Result<(), BoxError>` you can replace the `.unwrap()` with `?`

Reply:
Nice! Hadnt done this before